### PR TITLE
bpo-37013: Fix the error handling in socket.if_indextoname()

### PR DESCRIPTION
**Pull request by ZackerySpytz**
https://bugs.python.org/issue37013

### CODE REVIEW

**Comment by ZackerySpytz** [comment]
I don't think this needs a unit test, mainly because `(unsigned long)-1` can vary across platforms.

**Review by matrixise — APPROVED** [review]
Thank you for your contribution.

**Comment by zooba** [comment]
@matrixise I thought we gave you merge permissions? Want to hit the button? :)

**Review by serhiy-storchaka — COMMENTED** [review]
`if_indextoname()` takes `unsigned int`. Therefore highest bits of `index` can be ignored.

If UINT_MAX < index <= ULONG_MAX -- ignore highest bits.
If ULONG_MAX < index -- OverflowError.

I suggest to handle large index uniformly.
```c
if ((unsigned long)(unsigned int)index != index) {
    PyErr_SetString(OverflowError, "...");
}
```

**Comment by ned-deily** [comment]
This PR seems to have stalled.  @serhiy-storchaka, are you waiting for @ZackerySpytz to address your comments?

**Comment by serhiy-storchaka** [comment]
Yes, I am waiting for addressing my comments.

And please add tests for `socket.if_indextoname()` called with `2**32-1`, `2**32`, `2**64-1`, `2**64`. OSError and OverflowError should be ignored, but the test should not crash.

**Comment by matrixise** [comment]
> @matrixise I thought we gave you merge permissions? Want to hit the button? :)

@zoobar yep, with great powers, great responsibilities, but I don't want to crash Python ;-)

@ZackerySpytz could you update your PR with the comments of @serhiy-storchaka ? Thank you

**Comment by csabella** [comment]
@ZackerySpytz, please address the code review comments.  Thanks!

**Comment by csabella** [comment]
@serhiy-storchaka, please re-review as I believe your comments have been addressed.  Thank you!

**Review by iritkatriel — APPROVED** [review]

**Review by iritkatriel — CHANGES_REQUESTED** [review]
@ZackerySpytz, ping re the code review comments. Thanks!

**Review by iritkatriel — APPROVED** [review]
Sorry about the noise, I believe the review comments have been addressed.

File: Modules/socketmodule.c
Comment (vstinner) [inline Modules/socketmodule.c]: I would prefer to add an ``unsigned long index_long;`` variable and use it with PyLong_AsUnsignedLong(), and then downcast it into the ``index`` variable of unsigned int or NET_IFINDEX type, and check for overflow by comparing the two variables.

I prefer to call if_indextoname() with the expected type, with the `index` variable.

**Comment by ambv** [comment]
This missed the boat for inclusion in Python 3.9 which accepts security fixes only as of today.

**Comment by ambv** [comment]
Actually, wait, nevermind. This is a crasher and crashers are security issues. This **should** go into 3.7 - 3.9 once it lands.

**Comment by ambv** [comment]
I rebased and made changes per @vstinner's suggestions.

File: Modules/socketmodule.c
Comment (vstinner) [inline Modules/socketmodule.c:7085]: On Windows, NET_IFINDEX type is unsigned long, and so this change is not correct:
https://github.com/wine-mirror/wine/blob/11cb3bbf6908261458e8b6c0ad44f7daab8b9266/include/ifdef.h#L26

File: Modules/socketmodule.c
Comment (serhiy-storchaka) [inline Modules/socketmodule.c]: It should always be `unsigned long`.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]

File: Lib/test/test_socket.py
Comment (serhiy-storchaka) [inline Lib/test/test_socket.py]: Unfortunately this test does not work. It is passed even with not fixed code.

File: Modules/socketmodule.c
Comment (serhiy-storchaka) [reply Modules/socketmodule.c:7085]: This change does not have effect on Windows, but if fixes integer overflow on other platforms, where the argument is `unsigned int`.

File: Modules/socketmodule.c
Comment (vstinner) [inline Modules/socketmodule.c]: I suggest to move this variable just before if_indextoname() call, and write `unsigned long index_long = ...;` if it fits into 80 columns.

**Review by vstinner — APPROVED** [review]
LGTM

File: Lib/test/test_socket.py
Comment (serhiy-storchaka) [reply Lib/test/test_socket.py]: Currently it crashes the unfixed code, so I restored it.

File: Modules/socketmodule.c
Comment (serhiy-storchaka) [inline Modules/socketmodule.c:7082]: And the main change is that it checks PyErr_Occurred() now.

**Review by serhiy-storchaka — APPROVED** [review]

**Pull request merged**
